File: quick/golang/quickNums.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)


func QuickSort(values []int, began, end int, comp, swap *int) {
	i := began
	j := end -1
	pivo := values[(began + end) / 2]

	for i <= j {
		for values[i] < pivo && i < end {
			i++
			*comp+=2
		}
		for values[j] > pivo && j > began {
			j--
			*comp+=2
		}
		if i<=j {
			values[i], values[j] = values[j], values[i]
			*swap++
			i++
			j--
		}
	}
	if j > began {
		QuickSort(values, began, j+1, comp, swap)
	}
	if i < end {
		QuickSort(values, i, end, comp, swap)
	}
}


func main() {
	input, err := os.Open("numeros 1.txt")
	if err != nil {
		panic(err)
	}
	defer input.Close()
	output, err := os.OpenFile("./quickGoNums.txt", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	defer input.Close()

	vetor := make([]int, 0)
	var comp, swap int
	num, err := strconv.Atoi(os.Args[1])
	for i := 0; i < num; i++ {
		var aux int
		fmt.Fscan(input, &aux)
		vetor = append(vetor, aux)
	}


	inicio := time.Now()
	QuickSort(vetor, 0, num, &comp, &swap)
	fim := time.Now()


	fmt.Fprintf(output, "%s,%v,%d,%d\n", os.Args[1], float64(fim.Sub(inicio).Seconds()), comp, swap)

}

File: quick/cpp/quickNums.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <time.h>
using namespace std;

template <typename T>
void quicksort(vector<T> &values, int began, int end, unsigned long &comp, unsigned long &swap)
{
    int i, j;
    T pivo, aux;
    i = began;
    j = end - 1;
    pivo = values[(began + end) / 2];
    while (i <= j)
    {
        while (values[i] < pivo && i < end)
        {
            i++;
            comp+=2;
        }
        while (values[j] > pivo && j > began)
        {
            j--;
        comp+=2;
        }
        if (i <= j)
        {
            aux = values[i];
            values[i] = values[j];
            values[j] = aux;
            swap++;
            i++;
            j--;
        }
    }
    if (j > began)
        quicksort(values, began, j + 1, comp, swap);
    if (i < end)
        quicksort(values, i, end, comp, swap);
}

int main(int argc, char *argv[2])
{
    ifstream input("numeros 1.txt");
    ofstream output("./quickCppNums.txt", ios_base::app);
    vector<int> vetor;
    int aux;
    unsigned long comp = 0, swap = 0;
    clock_t inicio, fim;

    for (int i = 0; i < atoi(argv[1]); i++)
    {
        input >> aux;
        vetor.push_back(aux);
    }

    inicio = clock();

    quicksort(vetor, 0, size(vetor), comp, swap);

    fim = clock();

    output << argv[1] << "," << (float)(((fim - inicio) + 0.0) / CLOCKS_PER_SEC) << "," << comp << "," << swap << "\n";

    return 0;
}